### PR TITLE
Fix #3031: incompatibility with LaTeX package ``tocloft``

### LINKED ISSUE #3031: incompatibility with LaTeX package ``tocloft``
Using `tocloft` package to customize the table of contents (for example to increase the size for the page numbers, which is not big enough for some fonts and page numbers exceeding 100) leads to an error with the `manual`  type of document.

The reason is that `tocloft` redefines `\tableofcontents` at the location of `\begin{document}`, and the `\pagenumbering` which is modified by `sphinxmanual.cls` is not reset. Hence letters are used for the page numbers, and if the document contains more than 26 pages, there will be errors.

The fix: Sphinx should use `\sphinxtableofcontents` as the name of its modified code for `\tableofcontents`. This should be done independently of the issue with `tocloft`, in order to facilitate in the future LaTeX styling for Sphinx users.

### PR DESCRIPTION
**Pull request by jfbu**
Use `\sphinxtableofcontents`,  and environments `sphinxtheindex`
and `sphinxthebibliography`.

Similarly  the redefined `\maketitle` should be called `\sphinxmaketitle`.

### CODE REVIEW

File: doc/config.rst [hunk 1]
```diff
@@ -1739,13 +1739,17 @@ These options influence LaTeX output. See further :doc:`latex`.
...
+        redefined by the Sphinx ``manual`` and ``howto`` classes.) Override
+        if you want to
         generate a differently-styled title page.
      ``'releasename'``
         value that prefixes ``'release'`` element on title page, default
         ``'Release'``.
      ``'tableofcontents'``
-        "tableofcontents" call, default ``'\\tableofcontents'``.  Override if
+        "tableofcontents" call, default ``'\\sphinxtableofcontents'`` (it is a
+        wrapper of unmodified ``tableofcontents``.)
```
Comment (jfbu) [inline doc/config.rst]: typo here `\\tableofcontents`

File: sphinx/texinputs/sphinxhowto.cls [hunk 2]
```diff
@@ -91,21 +90,18 @@
...
 % For an article document class this environment is a section,
 % so no page break before it.
-\let\py@OldThebibliography=\thebibliography
-\renewcommand{\thebibliography}[1]{
+\newenvironment{sphinxthebibliography}[1]{%
   \phantomsection
-  \py@OldThebibliography{1}
-  \addcontentsline{toc}{section}{\bibname}
-}
+  \begin{thebibliography}{\detokenize{#1}}%
```
Comment (jfbu) [inline sphinx/texinputs/sphinxhowto.cls:95]: sometimes `#1` can be something like `Ref_1`, but the underscore causes a problem. With `\detokenize` we silence such problems. It is better than just using `1` which just doesn't take into account the size of the widest label.

File: sphinx/texinputs/sphinxhowto.cls [hunk 3]
```diff
@@ -91,21 +90,18 @@
...
+  \addcontentsline{toc}{chapter}{\bibname}}{\end{thebibliography}}
```
Comment (jfbu) [inline sphinx/texinputs/sphinxhowto.cls]: sorry chapter should be section, copy-paste error. Will fix in next commit.

File: sphinx/texinputs/sphinxhowto.cls [hunk 4]
```diff
@@ -91,21 +90,18 @@
...
+\@ifclassloaded{memoir}
+ {\newenvironment{sphinxtheindex}{\begin{theindex}}{\end{theindex}}}
+ {\newenvironment{sphinxtheindex}{%
     \phantomsection
-    \py@OldTheindex
-    \addcontentsline{toc}{section}{\indexname}
-  }
-}
+    \begin{theindex}%
+    \addcontentsline{toc}{chapter}{\indexname}}{\end{theindex}}}
```
Comment (jfbu) [inline sphinx/texinputs/sphinxhowto.cls]: here too chapter should be section, sorry. Will fix.

**Comment by tk0miya** [comment]
LGTM

**Comment by jfbu** [comment]
Thanks @tk0miya ! I will merge now and then update CHANGES.

**Pull request merged**